src/menus/EditMenus.cpp

```diff
@@ -138,9 +139,16 @@ void DoPasteNothingSelected(AudacityProject &project, const TrackList& src, doub
    // Select some pasted samples, which is probably impossible to get right
    // with various project and track sample rates.
    // So do it at the sample rate of the project
-   double projRate = ProjectRate::Get( project ).GetRate();
-   double quantT0 = QUANTIZED_TIME(t0, projRate);
-   double quantT1 = QUANTIZED_TIME(t1, projRate);
+   const double projRate = ProjectRate::Get( project ).GetRate();
+   const double projTempo = ProjectTimeSignature::Get(project).GetTempo();
+   const double srcTempo =
+      pFirstNewTrack ? pFirstNewTrack->GetProjectTempo().value_or(projTempo) :
+                       projTempo;
+   // Apply adequate stretching to the selection. A selection of 10 seconds of
+   // audio in project A should become 5 seconds in project B if tempo in B is
+   // twice as fast.
+   const double quantT0 = QUANTIZED_TIME(t0 * srcTempo / projTempo, projRate);
+   const double quantT1 = QUANTIZED_TIME(t1 * srcTempo / projTempo, projRate);
```

Comment:
The commit comment "Preserve beat length of selection when copying across projects" describes this part of it.  Perhaps the rest of this commit needs other commit comments, whether in the same or a separate commit.